### ISSUE TITLE
Implement uiTable

### ISSUE THREAD

**Issue by mischnic** [parro-it/libui-napi#45]
🙈

**Comment by parro-it** [comment]
Good news! I'm invacation rightnow. Will be back on end of august, I will start toadd tables to bindings.

**Comment by parro-it** [comment]
I'm back on the road again, and I started to work on the new implementation. As a first step, I did a PR where I changed the download process of libui binaries in order to use @andlabs repo

**Comment by mischnic** [comment]
libui issue: https://github.com/andlabs/libui/issues/425

**Comment by mischnic** [comment]
How is the progress? Did you implement all libui table features? 

---

Regarding this todo:

https://github.com/parro-it/libui-napi/blob/e3848cf61e1dd332ee9ce3eb0765a101c89877c2/src/image.c#L8-L11

I guess the only way is reference counting.

**Comment by andlabs** [comment]
uiTable does not copy images; they are used as soon as you give it to them, and then forgotten about.

**Comment by mischnic** [comment]
That's true. The images are only used in the `modelCellValue` callback, so when the JS object get's garbage collected, the image can be deleted (no reference is stored inside native code).

**Comment by andlabs** [comment]
Oh. I thought I documented the guarantee on lifetime, but basically by the top of the next main loop iteration uiTable will have finished using it. I'm not sure whether creating a new uiImage each time you extract data is a good idea, but eh.

**Comment by mischnic** [comment]
> I'm not sure whether creating a new uiImage each time you extract data is a good idea, but eh.

What do you mean by extracting data? There is a javascript image wrapper which forwards the calls to the actual uiImage and when the wrapper would get gc'ed, the uiImage is freed as well. This underlying uiImage also gets passed to libui in the table callback.

**Comment by parro-it** [comment]
> How is the progress? Did you implement all libui table features?

Almost, I have to finish the abstraction layer on top of models.
But basic libui methods binding should be all there. There are some memory management to improve on.... one is the one you discovered here, the other is in order to keep model instances alive while they are used by any or more tables.

> I guess the only way is reference counting when using an image in a table.

I was thinking the same, we can increment the reference when an image is returned by the model handler, and someway decrements it in next loop step, as @andlabs suggested... 

Another strangeness to solve is here: https://github.com/parro-it/libui-napi/blob/e3848cf61e1dd332ee9ce3eb0765a101c89877c2/src/table-model.c#L234
AFAIK that string should be freed immediately, but if I do it there, the program SEGFAULT.

**Comment by mischnic** [comment]
>  AFAIK that string should be freed immediately, but if I do it there, the program SEGFAULT.

Doesn't segfault for me, but according to ui.h:

> uiTableValueString() returns the string stored in v. The returned string is owned by v.

So it's not copied and shouldn't get freed.

**Comment by jbrodriguez** [comment]
I'm about to build a desktop app and deciding wether to use electron or proton-native.

The only thing missing from a workable proton-native poc is lists/tables (as mentioned in https://github.com/kusti8/proton-native/issues/99)

So I was wondering when do guys think this could become available ?

Early next year ? Or before the end of the year ?

**Comment by parro-it** [comment]
@jboodriguez : I definitely think before the end of the year.  

Anyway, please be aware that I'm doing this in my spare time, for free.
So it greately depends on how I'll be tired/motivated/busy on other stuff. 

Don't take that date as strict deadline, only as an informed guess...

**Comment by benbucksch** [comment]
If this helps as motivation: I'm trying out Proton Native, and without a table widget with selection (single and multi), I have no chance to build my app.

**Comment by benbucksch** [comment]
@parro-it : Here's how an email client without tables looks like: :-)
![image](https://user-images.githubusercontent.com/1907525/70370747-5dd6b000-18cb-11ea-9578-5bc44a384316.png)
